lsp/nls/src/requests/hover.rs

```diff
@@ -1,68 +1,169 @@
+            ret.values = values;
+            ret.metadata = metadata;
+        } else if def.path.is_empty() {
+            ret.values.extend(def.value.iter().cloned());
+        }
+    }
+
+    Some(ret)
+}
+
+fn term_hover(rt: &RichTerm, server: &Server) -> Option<HoverData> {
+    let ty = server.lin_registry.get_type(rt).cloned();
+    let span = rt.pos.into_opt()?;
+
+    match rt.as_ref() {
+        Term::Op1(UnaryOp::StaticAccess(id), parent) => {
+            let resolver = FieldResolver::new(server);
+            let parents = resolver.resolve_term(parent);
+            let (values, metadata) = values_and_metadata_from_field(parents, id.ident());
+            Some(HoverData {
+                values,
+                metadata,
+                span,
+                ty,
+            })
+        }
+        _ => Some(HoverData {
+            values: vec![rt.clone()],
+            metadata: vec![],
+            span,
+            ty,
+        }),
+    }
+}
+
 pub fn handle(
     params: HoverParams,
-    id: RequestId,
+    req_id: RequestId,
     server: &mut Server,
 ) -> Result<(), ResponseError> {
     let pos = server
         .cache
         .position(&params.text_document_position_params)?;
 
-    debug!("pos of hovered item: {pos:?}");
+    let hover_data = server
+        .lookup_ident_by_position(pos)?
+        .and_then(|ident| ident_hover(ident, server));
 
-    let linearization = server.lin_cache_get(&pos.src_id)?;
-    let item = linearization.item_at(pos);
+    let hover_data = match hover_data {
+        Some(h) => Some(h),
+        None => server
+            .lookup_term_by_position(pos)?
+            .and_then(|rt| term_hover(rt, server)),
+    };
 
-    Trace::enrich(&id, linearization);
+    if let Some(hover) = hover_data {
+        let mut contents = Vec::new();
 
-    if item.is_none() {
-        server.reply(Response::new_ok(id, Value::Null));
-        return Ok(());
-    }
+        if let Some(ty) = hover.ty {
+            contents.push(nickel_string(ty.to_string()));
+        } else {
+            // Unclear whether it's useful to report `Dyn` all the
+            // time, but it matches the old behavior.
+            contents.push(nickel_string("Dyn".to_string()));
+        }
+
+        let mut contracts: Vec<_> = hover
+            .metadata
+            .iter()
+            .flat_map(|m| &m.annotation.contracts)
+            .chain(hover.values.iter().flat_map(annotated_contracts))
+            .map(|contract| contract.label.typ.to_string())
+            .collect();
 
-    let item = item.unwrap().to_owned();
+        contracts.sort();
+        contracts.dedup();
 
-    debug!("Found item {:?}", item);
+        contents.extend(contracts.into_iter().map(nickel_string));
 
-    let (ty, meta) = linearization.get_type_and_metadata(&item, &server.lin_registry);
+        // Not sure how to do documentation merging yet, so pick the first non-empty one.
```

Comment:
I think it's the current behavior of merging as well. We should probably abstract that into Nickel (maybe define `struct Doc(Rc<String>)` and implement `Combine` on it, but I reckon it's not yet the case)